### ISSUE TITLE
bfloat16 overflow during training session

### ISSUE THREAD

**Issue by huseinzol05** [state-spaces/mamba#6]
1. I tried vanilla pytorch training loop using bfloat16, the loss got overflow, https://github.com/mesolitica/malaya/blob/5.1/pretrained-model/mamba/causallm-130m-bf16.ipynb
2. so I tried vanilla pytorch training loop using fp32, the loss is ok, https://github.com/mesolitica/malaya/blob/5.1/pretrained-model/mamba/causallm-130m-fp32.ipynb
3. I thought maybe because no gradient clipping and etc, so I tried using HuggingFace trainer with Deepspeed, https://github.com/mesolitica/malaya/blob/5.1/pretrained-model/mamba/causallm-130m-trainer-deepspeed-bf16.ipynb, the loss got overflow.
4. so I removed deepspeed and use fp32 in HuggingFace trainer, https://github.com/mesolitica/malaya/blob/5.1/pretrained-model/mamba/causallm-130m-trainer-fp32.ipynb, the loss is ok.

If bfloat16 is not working, deepspeed is not going to work.

**Comment by tridao** [comment]
We don't use deepspeed, just Pytorch AMP (bf16) to train models. Can you try that?

**Comment by tridao** [comment]
Model parameters should be in fp32, just like in Pytorch [AMP](https://pytorch.org/docs/stable/amp.html) docs.

**Comment by huseinzol05** [comment]
Looks good on amp bf16, https://github.com/mesolitica/malaya/blob/5.1/pretrained-model/mamba/causallm-130m-trainer-bf16.ipynb, and I saw there is 2.8B checkpoint, what is the batch size and it use 2k context length right based on the paper? Is it the 2.8B model fit in a single A100 80GB?

**Comment by tridao** [comment]
The 2.8B uses total batch size of 1M tokens (following GPT3 paper), and seqlen=2k. Activation memory should be around the same as an optimized transformer (e.g. with FlashAttention), so it should fit in a single A100 80GB (you might need to adjust gradient accumulation to fit).
For models around 1B-3B on 8xA100s, sharding the optimizer states (e.g. with Pytorch distributed [optimizer](https://pytorch.org/docs/stable/distributed.optim.html), equivalent to ZeRO-stage1) will help reduce the amount of memory needed.

**Comment by huseinzol05** [comment]
Thanks!

**Comment by huseinzol05** [comment]
I tried with Zero2 FP32, sometime not overflow, sometime overflow, after that tried 1e-6 lr and 0.1 gradient norm, no longer overflow but 0.1 gradient norm is too low for pretraining.

**Comment by albertfgu** [comment]
Thanks for your explorations! I added a little warning to the README about the dtype, but it's quite useful for people to post their observations here. We've actually never seen these types of instabilities during training; as Tri said, we just use native PyTorch AMP.

**Comment by binxuan** [comment]
I am using pytorch's FSDP with bf16 for training. Looks like I encountered similar issue with NaN loss.

**Comment by geronimi73** [comment]
> I am using pytorch's FSDP

what is your `auto_wrap_policy`? if you don't mind sharing it
@binxuan

**Comment by btrude** [comment]
> > I am using pytorch's FSDP
> 
> what is your `auto_wrap_policy`? if you don't mind sharing it @binxuan

This worked for me:

```py
from mamba_ssm.modules.mamba_simple import Block
from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy

auto_wrap_policy = partial(
    transformer_auto_wrap_policy,
    transformer_layer_cls={Block,},
)
```

I was also able to pretrain a 200m mamba lm on 12b tokens yesterday in bf16 without issue. Testing fsdp + bf16 right now and everything works as expected as well.

**Comment by lwang2070** [comment]
Hi, I am using pytorch_lighting trainer with [bf16-mixed](https://lightning.ai/docs/pytorch/stable/common/precision_intermediate.html#bfloat16-mixed-precision) precision, i.e., model params in f32 and bf16 to train the model (pytorch_lightning also use Pytorch AMP under the hood), to train Mamba on DNA data, however, the training is still 
quite unstable

![Loss on hg38](https://github.com/state-spaces/mamba/assets/94618763/0c52f7ee-a789-4551-ad39-2c178ad9f673)

Weirdly, I trained using the same code and same model (different size) with f16-mixed precision on NLP data (wiki103), and such instability did not occur. Any suggestion?

**Comment by gpantaz** [comment]
> Hi, I am using pytorch_lighting trainer with [bf16-mixed](https://lightning.ai/docs/pytorch/stable/common/precision_intermediate.html#bfloat16-mixed-precision) precision, i.e., model params in f32 and bf16 to train the model (pytorch_lightning also use Pytorch AMP under the hood), to train Mamba on DNA data, however, the training is still quite unstable
> 
> ![Loss on hg38](https://private-user-images.githubusercontent.com/94618763/297616938-0c52f7ee-a789-4551-ad39-2c178ad9f673.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1Njk0ODksIm5iZiI6MTcwNTU2OTE4OSwicGF0aCI6Ii85NDYxODc2My8yOTc2MTY5MzgtMGM1MmY3ZWUtYTc4OS00NTUxLWFkMzktMmMxNzhhZDlmNjczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDA5MTMwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY3MmFmZTY0MTViZGNkMDZkNjZlYTg4OTMzNzE4YTcwZGZlMTA0ZjhlYTZhYTM0YzlmNTQ1OWNiMDc1MDY1YjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wwrM1A5Da99l1P2i-svSlUGZHKgqxwsEYhlCedln1SU)
> 
> Weirdly, I trained using the same code and same model (different size) with f16-mixed precision on NLP data (wiki103), and such instability did not occur. Any suggestion?

I also have observed similar instabilities with using PL trainer. In my case I am creating custom embeddings which I concantenate along with the text embeddings from mamba before feeding them to the mixer model. I have tried lowering the learning rate, disabling AMP but my loss goes to nan values.

**Comment by sustcsonglin** [comment]
> I also have observed similar instabilities with using PL trainer. In my case I am creating custom embeddings which I concantenate along with the text embeddings from mamba before feeding them to the mixer model. I have tried lowering the learning rate, disabling AMP but my loss goes to nan values.

I have similar issues.

**Comment by apoorv2904** [comment]
@gpantaz  and @lwang2070  were you able to fix the issue of bf16 or fp16 training with PL? I also see training issues for my model.

**Comment by gpantaz** [comment]
No sadly :/

**Comment by layumi** [comment]
Similar issue here.  I am using pytorch_lighting trainer with [bf16-mixed] precision + ddp_strategy ="deepspeed_stage_2".

The same code with same hyper-parameter is Okay on A100-SXM4-80GB even I tune larger learning rate, 
but failed on RTX A6000 and 4090.
